server/audit_math/supersimple.py

```diff
@@ -0,0 +1,172 @@
+    obs_u2 = sample_results["2-under"]
+    num_sampled = sample_results["sample_size"]
+
+    if num_sampled:
+        r1 = obs_o1 / num_sampled
+        r2 = obs_o2 / num_sampled
+        s1 = obs_u1 / num_sampled
+        s2 = obs_u2 / num_sampled
+    else:
+        r1 = o1
+        r2 = o2
+        s1 = u1
+        s2 = u2
+
+    denom = (
+        math.log(1 - contest.diluted_margin / (2 * gamma))
+        - r1 * math.log(1 - 1 / (2 * gamma))
+        - r2 * math.log(1 - 1 / gamma)
+        - s1 * math.log(1 + 1 / (2 * gamma))
+        - s2 * math.log(1 + 1 / gamma)
+    )
+
+    if denom >= 0:
+        return contest.ballots
+
+    n0 = math.ceil(math.log(risk_limit) / denom)
+
+    # Round up one-vote differences.
+    r1 = math.ceil(r1 * n0)
+    s1 = math.ceil(s1 * n0)
+
+    return nMin(risk_limit, contest, r1, r2, s1, s2)
+
+
+def compute_risk(risk_limit, contest, cvrs, sample_cvr):
+    """
+    Computes the risk-value of <sample_results> based on results in <contest>.
+
+    Inputs:
+        contests       - the contests and results being audited
+        cvrs           - mapping of ballot_id to votes:
+                {
+                    'ballot_id': {
+                        'contest': {
+                            'candidate1': 1,
+                            'candidate2': 0,
+                            ...
+                        }
+                    ...
+                }
+
+        sample_cvr - the CVR of the audited ballot
+                {
+                    'ballot_id': {
+                        'contest': {
+                            'candidate1': 1,
+                            'candidate2': 0,
+                            ...
+                        }
+                    ...
+                }
+
+    Outputs:
+        measurements    - the p-value of the hypotheses that the election
+                          result is correct based on the sample, for each winner-loser pair.
+        confirmed       - a boolean indicating whether the audit can stop
+    """
+
+    p = 1
+
+    V = contest.diluted_margin * len(cvrs)
+
+    U = 2 * gamma / contest.diluted_margin
+
+    result = False
+    for ballot in sample_cvr:
+        e_r = 0
+
+        if contest.name not in sample_cvr[ballot]:
+            continue
+        for winner in contest.winners:
+            for loser in contest.losers:
+                v_w = cvrs[ballot][contest.name][winner]
+                a_w = sample_cvr[ballot][contest.name][winner]
+
+                v_l = cvrs[ballot][contest.name][loser]
+                a_l = sample_cvr[ballot][contest.name][loser]
+
+                V_wl = contest.candidates[winner] - contest.candidates[loser]
+
+                e = ((v_w - a_w) - (v_l - a_l)) / V_wl
+                if e > e_r:
+                    e_r = e
+
+        denom = (2 * gamma) / V
+        p_b = (1 - 1 / U) / (1 - (e_r / denom))
+        p *= p_b
+
+    if 0 < p < risk_limit:
+        result = True
+
+    return p, result
```

Comment:
curious why this is called `supersimple` lol. is there a more complex method coming? wondering if it would be helpful to have "ballot comparison" in the name

Reply:
It's called that because that's what Stark called it. "Ballot comparison" is fairly vague and can refer to many different things, unfortunately. FWIW, we don't use "ballot polling" or "batch comparison" in our other modules for this reason too...